nway/image_processing_utils.py

```diff
@@ -0,0 +1,145 @@
+        3 x 3 transformation matrix
+
+    """
+
+    cvmotion = {
+            "MOTION_TRANSLATION": cv2.MOTION_TRANSLATION,
+            "MOTION_EUCLIDEAN": cv2.MOTION_EUCLIDEAN,
+            "MOTION_AFFINE": cv2.MOTION_AFFINE,
+            "MOTION_HOMOGRAPHY": cv2.MOTION_HOMOGRAPHY}
+
+    # Define termination criteria
+    criteria = (
+            cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT,
+            maxCount,
+            epsilon)
+
+    dx = dy = 0.0
+    if preregister:
+        (dx, dy), _ = cv2.phaseCorrelate(img_fixed.astype('float32'),
+                                         img_moving.astype('float32'))
+
+    tform = np.array([[1.0, 0.0, dx],
+                      [0.0, 1.0, dy]]).astype('float32')
+    if motion_type == 'MOTION_HOMOGRAPHY':
+        hrow = np.array([0.0, 0.0, 1.0]).astype('float32')
+        tform = np.vstack((tform, hrow))
+
+    try:
+        # Run the ECC algorithm. The results are stored in warp_matrix.
+        ccval, tform = cv2.findTransformECC(
+                img_fixed,
+                img_moving,
+                tform,
+                cvmotion[motion_type],
+                criteria,
+                None,
+                gaussFiltSize)
+    except cv2.error:
+        logger.error("failed to align images.")
+        raise
+
+    # not all the transforms output 3 x 3
+    if tform.shape == (2, 3):
+        tform = np.vstack((tform, [0, 0, 1]))
+
+    return tform
+
+
+def contrast_adjust(image, CLAHE_grid, CLAHE_clip):
+    """contrast adjust images
+
+    Parameters
+    ----------
+    image: numpy.ndaray
+        image to pre-process. CV_8UC1 (uint8) or CV_16UC1 (uint16)
+    CLAHE_grid : int
+        passed as tileGridSize to cv2.createCLAHE. If -1, skipped
+    CLAHE_clip : float
+        passed as clipLimit to cv2.createCLAHE
+
+    Returns
+    -------
+    image: numpy.narray
+        contrast-adjusted image
+
+    """
+    if CLAHE_grid != -1:
+        clahe = cv2.createCLAHE(
+            clipLimit=CLAHE_clip,
+            tileGridSize=(CLAHE_grid, CLAHE_grid))
+        image = clahe.apply(image)
+    return image
+
+
+def warp_image(image, tform, motion_type, dst_shape):
+    """
+
+    Parameters
+    ----------
+    image: numpy.ndarray
+        image to warp. Assumes uint8
+    tform: numpy.ndarray
+        3x3 transformation matrix
+    motion_type: str
+        kind of motion. See function `register_image_pair` for possible
+        values
+    dst_shape: tuple
+        desired shape of warped image
+
+    Returns
+    -------
+    image: numpy.ndarray
+        warped image, forced to uint8
+
+    """
+    if motion_type == 'MOTION_HOMOGRAPHY':
+        warp = cv2.warpPerspective
+    else:
+        warp = cv2.warpAffine
+        if tform.shape[0] == 3:
+            tform = tform[0:2]
+
+    image = warp(
+            image,
+            tform,
+            dst_shape[::-1],
+            flags=cv2.INTER_LINEAR + cv2.WARP_INVERSE_MAP).astype(np.uint8)
```

Comment:
Just to double check but 2P mesoscope images use `uint8`?

Some microscopy systems take `uint16` images and casting to `uint8` could result in bad clipping.

Reply:
this is registration on average projection images, not images right out of the scope. They are 8-bit.
```
identify /allen/programs/braintv/production/visualbehavior/prod0/specimen_843122504/ophys_session_908569379/ophys_experiment_909184300/processed/ophys_cell_segmentation_run_909926779/avgInt_a1X.png
/allen/programs/braintv/production/visualbehavior/prod0/specimen_843122504/ophys_session_908569379/ophys_experiment_909184300/processed/ophys_cell_segmentation_run_909926779/avgInt_a1X.png PNG 512x512 512x512+0+0 8-bit sRGB 256c 72.5KB 0.000u 0:00.009
```